File: NYCGayBars/Logic/Geo.swift
import Foundation

/// Geolocation helpers ported from RN lib/geo.ts.
enum Geo {
    private static let milesPerDegLat = 69.0

    /// Squared equirectangular distance — fine for ranking, not metric.
    private static func sqDist(_ lat: Double, _ lng: Double, _ bar: Bar, _ latRadCos: Double) -> Double {
        let dLat = bar.latitude - lat
        let dLon = (bar.longitude - lng) * latRadCos
        return dLat * dLat + dLon * dLon
    }

    struct Bounds { var minLat, maxLat, minLon, maxLon: Double }

    /// Bounding box of each neighborhood's bars, computed once from BARS.
    static let neighborhoodBounds: [String: Bounds] = {
        var bounds: [String: Bounds] = [:]
        for bar in AppData.bars {
            if var b = bounds[bar.neighborhood] {
                b.minLat = min(b.minLat, bar.latitude)
                b.maxLat = max(b.maxLat, bar.latitude)
                b.minLon = min(b.minLon, bar.longitude)
                b.maxLon = max(b.maxLon, bar.longitude)
                bounds[bar.neighborhood] = b
            } else {
                bounds[bar.neighborhood] = Bounds(
                    minLat: bar.latitude, maxLat: bar.latitude,
                    minLon: bar.longitude, maxLon: bar.longitude)
            }
        }
        return bounds
    }()

    /// How many neighborhoods sit entirely inside the visible map region.
    static func fullyVisibleNeighborhoods(_ region: Region) -> Int {
        let minLat = region.latitude - region.latitudeDelta / 2
        let maxLat = region.latitude + region.latitudeDelta / 2
        let minLon = region.longitude - region.longitudeDelta / 2
        let maxLon = region.longitude + region.longitudeDelta / 2
        var count = 0
        for b in neighborhoodBounds.values {
            if b.minLat >= minLat && b.maxLat <= maxLat && b.minLon >= minLon && b.maxLon <= maxLon {
                count += 1
            }
        }
        return count
    }

    /// Equirectangular distance in miles — accurate enough at city scale.
    static func distanceMiles(_ lat: Double, _ lng: Double, _ bar: Bar) -> Double {
        let cos = Foundation.cos(lat * .pi / 180)
        return sqrt(sqDist(lat, lng, bar, cos)) * milesPerDegLat
    }

    /// The bar closest to the given coordinates.
    static func nearestBar(_ latitude: Double, _ longitude: Double) -> Bar {
        let cos = Foundation.cos(latitude * .pi / 180)
        var best = AppData.bars[0]
        var bestDist = Double.infinity
        for bar in AppData.bars {
            let dist = sqDist(latitude, longitude, bar, cos)
            if dist < bestDist { bestDist = dist; best = bar }
        }
        return best
    }

    /// Neighborhoods ordered by how close their nearest bar is (closest first).
    static func neighborhoodsByProximity(_ latitude: Double, _ longitude: Double) -> [String] {
        let cos = Foundation.cos(latitude * .pi / 180)
        var minDist: [String: Double] = [:]
        for bar in AppData.bars {
            let dist = sqDist(latitude, longitude, bar, cos)
            if let cur = minDist[bar.neighborhood] {
                if dist < cur { minDist[bar.neighborhood] = dist }
            } else {
                minDist[bar.neighborhood] = dist
            }
        }
        return AppData.neighborhoods.sorted {
            (minDist[$0] ?? .infinity) < (minDist[$1] ?? .infinity)
        }
    }

    /// True when the string is a 5-digit ZIP code in our NYC dataset.
    static func isKnownZip(_ query: String) -> Bool {
        guard query.count == 5, query.allSatisfy({ $0.isNumber }) else { return false }
        return AppData.zipCentroids[query] != nil
    }

    /// The bar neighborhood closest to a NYC ZIP's centroid, or nil if unknown.
    static func neighborhoodForZip(_ zip: String) -> String? {
        guard isKnownZip(zip), let c = AppData.zipCentroids[zip] else { return nil }
        return nearestBar(c.lat, c.lng).neighborhood
    }
}

File: legacy-expo/lib/geo.ts
import { BARS, NEIGHBORHOODS } from "./bars";
import type { Bar } from "./types";
import { ZIP_CENTROIDS } from "./zips";

/** Squared equirectangular distance — fine for comparing/ranking, not metric. */
function sqDist(
  lat: number,
  lng: number,
  bar: Bar,
  latRadCos: number,
): number {
  const dLat = bar.latitude - lat;
  const dLon = (bar.longitude - lng) * latRadCos;
  return dLat * dLat + dLon * dLon;
}

const MILES_PER_DEG_LAT = 69.0;

type Bounds = { minLat: number; maxLat: number; minLon: number; maxLon: number };

/** Bounding box of each neighborhood's bars, computed once from BARS. */
const NEIGHBORHOOD_BOUNDS: Record<string, Bounds> = (() => {
  const bounds: Record<string, Bounds> = {};
  for (const bar of BARS) {
    const b = bounds[bar.neighborhood];
    if (!b) {
      bounds[bar.neighborhood] = {
        minLat: bar.latitude,
        maxLat: bar.latitude,
        minLon: bar.longitude,
        maxLon: bar.longitude,
      };
    } else {
      b.minLat = Math.min(b.minLat, bar.latitude);
      b.maxLat = Math.max(b.maxLat, bar.latitude);
      b.minLon = Math.min(b.minLon, bar.longitude);
      b.maxLon = Math.max(b.maxLon, bar.longitude);
    }
  }
  return bounds;
})();

type RegionLike = {
  latitude: number;
  longitude: number;
  latitudeDelta: number;
  longitudeDelta: number;
};

/** How many neighborhoods sit entirely inside the visible map region. */
export function fullyVisibleNeighborhoods(region: RegionLike): number {
  const minLat = region.latitude - region.latitudeDelta / 2;
  const maxLat = region.latitude + region.latitudeDelta / 2;
  const minLon = region.longitude - region.longitudeDelta / 2;
  const maxLon = region.longitude + region.longitudeDelta / 2;
  let count = 0;
  for (const b of Object.values(NEIGHBORHOOD_BOUNDS)) {
    if (
      b.minLat >= minLat &&
      b.maxLat <= maxLat &&
      b.minLon >= minLon &&
      b.maxLon <= maxLon
    ) {
      count++;
    }
  }
  return count;
}

/** Equirectangular distance in miles — accurate enough at city scale. */
export function distanceMiles(lat: number, lng: number, bar: Bar): number {
  const cos = Math.cos((lat * Math.PI) / 180);
  return Math.sqrt(sqDist(lat, lng, bar, cos)) * MILES_PER_DEG_LAT;
}

/** The bar closest to the given coordinates. */
export function nearestBar(latitude: number, longitude: number): Bar {
  const cos = Math.cos((latitude * Math.PI) / 180);
  let best = BARS[0];
  let bestDist = Infinity;
  for (const bar of BARS) {
    const dist = sqDist(latitude, longitude, bar, cos);
    if (dist < bestDist) {
      bestDist = dist;
      best = bar;
    }
  }
  return best;
}

/**
 * Neighborhoods ordered by how close their nearest bar is to the given
 * coordinates (closest first). Used to sort the filter chips by proximity.
 */
export function neighborhoodsByProximity(
  latitude: number,
  longitude: number,
): string[] {
  const cos = Math.cos((latitude * Math.PI) / 180);
  const minDist = new Map<string, number>();
  for (const bar of BARS) {
    const dist = sqDist(latitude, longitude, bar, cos);
    const cur = minDist.get(bar.neighborhood);
    if (cur === undefined || dist < cur) minDist.set(bar.neighborhood, dist);
  }
  return [...NEIGHBORHOODS].sort(
    (a, b) => (minDist.get(a) ?? Infinity) - (minDist.get(b) ?? Infinity),
  );
}

/**
 * True when the string is a 5-digit ZIP code in our NYC dataset. The regex
 * gate also keeps Object.prototype keys ("constructor" etc.) from passing
 * the `in` lookup.
 */
export function isKnownZip(query: string): boolean {
  return /^\d{5}$/.test(query) && query in ZIP_CENTROIDS;
}

/**
 * The bar neighborhood closest to a NYC ZIP code's centroid, or null if the
 * ZIP isn't in our dataset. Used to jump the neighborhood filter to a typed ZIP.
 */
export function neighborhoodForZip(zip: string): string | null {
  if (!isKnownZip(zip)) return null;
  const c = ZIP_CENTROIDS[zip];
  return nearestBar(c.lat, c.lng).neighborhood;
}
